statsmodels/tsa/base/tsa_model.py

```diff
@@ -319,9 +341,23 @@ def _get_index_loc(self, key, base_index=None):
                                             periods=len(index) + 1,
                                             freq=base_index.freq)
 
-        # Get the location (note that get_loc will throw a KeyError if key is
-        # invalid)
-        loc = index.get_loc(key)
+        # Get the location
+        if date_index:
+            # (note that get_loc will throw a KeyError if key is invalid)
+            loc = index.get_loc(key)
+        elif int_index or range_index:
+            # For Int64Index and RangeIndex, key is assumed to be the location
+            # and not an index value (this assumption is required to support
+            # RangeIndex)
+            try:
+                index[key]
+            # We want to raise a KeyError in this case, to keep the exception
+            # consistent across index types
+            except IndexError as e:
```

Comment:
a) why do we care that it is always a KeyError?  I'm not clear on why index.get_loc doesn't work in this case.
b) Personal preference for the "# We want to raise [...]" comment to go after the `except IndexError as e:`, with an extra level of indentation.  YMMV.
c) This method is pretty big, could stand to be refactored into smaller pieces.  This block (344-360) would be a nice scope because it (almost) just depends on `index` and `key`.  self-contained tests!

Reply:
First, to be honest, I do not fully understand the design decisions between `KeyError` and `IndexError` in Pandas. For example, the following:

```python
x = pd.Int64Index(np.arange(10))
x.get_loc(11)  # raises KeyError
x[11]          # raises IndexError
```

Now, this part of the code is relying on Pandas to tell me if the key is invalid (i.e. we don't want to have to figure it out ourselves), and we'd ideally like to check using `index.get_loc(key)` so that it works with date strings, etc. This also works fine for incrementing `Int64Index`es, but the problem is that for `RangeIndex`, the index values are integers too, so for example:

```python
x = pd.RangeIndex(start=0, stop=10, step=2)
print(x.get_loc(2))  # 1
print(x[2])          # 4
```

Now in this code I want to check that the key is a valid part of the index, so for a `RangeIndex` I have to do that using `index[key]` rather than `index.get_loc(key)` and for some reason this raises an `IndexError`. Intuitively, though, I'm searching for the existence of a key, and I want to raise a `KeyError` if it doesn't exist.

If there's a better way to do this check consistently, that's good too.

Reply:
(c) I suppose we could refactor it into `_get_index_loc` which then calls `_get_index_loc_int64index`, `_get_index_loc_rangeindex`, and `_get_index_loc_datetindex`. What do you think?

Reply:
I'll take a look at the KeyError vs IndexError issue.

On the splitting-up-big-methods, feel free to copy/paste from [here](https://github.com/jbrockmendel/sm2/blob/master/sm2/tsa/base/tsa_model.py).